ide/coqide/fileOps.ml

```diff
@@ -72,11 +72,11 @@ object(self)
       | Some f ->
         if not buffer#modified then do_revert f
         else
-          let answ = Configwin_ihm.question_box
+          let answ = Ui_dialogs.question_box
             ~title:"Modified buffer changed on disk"
-            ~buttons:["Revert from File";
-                      "Overwrite File";
-                      "Disable Auto Revert"]
+            ~buttons:[ButtonWithLabel "Revert from file";
+                      ButtonWithLabel "Overwrite file";
+                      ButtonWithLabel "Disable Auto Revert"]
```

Comment:
“Revert” means “forget the source code that is currently in the editor, and put the content of your Coq file (from which you got your Coq source or in which you last saved your Coq source) in the editor instead.” I’ve seen the feature evoked in the documentation along with the “External editor” feature: it is advised to enable auto revert if you use an external editor. I don’t really understand why they ask for a delay; when I tried to use it, CoqIDE seemed to ask me immediately when I activated the CoqIDE window if I wanted to revert, and I think most IDEs work like this. It is also common to have a “Revert” item in the File menu.

Reply:
A very poor choice of name.  The dictionary definition of "revert" is "return to (a previous state, practice, topic, etc.)".  "reload" is somewhat better.

Re-reading the file should be triggered by the CoqIDE window getting focus, but I'm not sure XWindows and/or GTK can do that.  (And writing should be triggered by CoqIDE losing focus.)

Reply:
> "reload" is somewhat better.

I completely agree. In fact, it also exists as a menu item:
https://garrigue.github.io/lablgtk/lablgtk3/lablgtk3/GtkStock/index.html#type-gtk_stock_id.REVERT_TO_SAVED
where “revert” generally corresponds to reloading a file which has not been changed since the last loading. In French, we also had the wrong word “Rétablir” (revert) instead of “Recharger” (reload).

I’ve just batched-replaced all occurrences of “revert” to “reload”.

Reply:
> Re-reading the file should be triggered by the CoqIDE window getting focus, but I'm not sure XWindows and/or GTK can do that.

It is a very common feature, so GTK or GDK probably have it, but lablgtk may not. We’ll see this later.

> (And writing should be triggered by CoqIDE losing focus.)

I don’t like this kind of automatic non-obvious behaviors. To me, the good practice is to get used to pressing Ctrl-S all the time :laughing:. I think such automatic behaviors lead people not to be aware of what their computer does, and (as said by many free software enthusiasts, but it may be even more important to me than to them) it is very bad. I also have a few practices which are not compatible with this feature, e.g. I will save something I want to test elsewhere, but before testing, I will change my buffer to what I want my code to be just after the testing (more concretely, this means I press: Ctrl-Z several times, Ctrl-S, Ctrl-Shift-Z until it has no effect, Alt-Tab, testing, Alt-Tab, Ctrl-S).

But obviously, I am not against providing this feature as a by-default-disabled preference.

Reply:
"obvious" is very subjective

Reply:
In fact, not so much. A computer behavior is obvious when it is a technical implication, a technical optimization, a way to avoid some side-effect or dependency, or at least quite close to some very well-known standard.
- For instance, saving a source file before compiling it is standard, technically simpler, avoids depending on another location on which the editor has the permission to write, and is also quite an optimization because when you spend time compiling files and testing the result, you usually do want the corresponding source files being saved. *But* I’ve already seen a software program (I don’t remember which) which does save source files in a temporary location (/tmp) for compilation, if they are not already saved (and then saving your source file requires a bit more computer work).
- On the other hand, saving when losing focus is not at all technically implied, may cause useless computer work, may have the side-effect of overwriting something that the user wanted to keep… And we could even take in account further implications of automatic saving, e.g. if you save on a distant location and have a limited Internet connection.

I won’t blame it if you like this kind of auto-saving feature (which usually has no wrong implication at all), but at first it didn’t exist and I think there are good reasons to avoid the magic as a default behavior.

Reply:
The IDE I've been using for the last 20 years (IntelliJ IDEA) has always autosaved when it loses focus and reloads when it gains focus.  It also has great diff capabilities, so I typically review my changes compared to git (in the IDE) before I commit.  I can edit the changes or revert individual changes in the diff display.  (Reviewing your own code is very helpful!)  Undesired changes are usually very easy to pick out.  I can also revert all changes to a set of files.

Explicitly saving means you need to think about the editor state and the file system state separately.  For editing, I've not found that a useful distinction and a drawback if you ever forget to save.  But it's a matter of personal preference.